### ISSUE TITLE
`...+variable ? arr1 : arr2` prints warning "Expected an identifier and instead saw '+'"

### ISSUE THREAD

**Issue by kolontsov** [jshint/jshint#3192]
jshint 2.9.5 prints warning `Expected an identifier and instead saw '+'` (E030) on this (or similar) code:

```console.log([...+process.env.TEST ? [1] : [], 2]);```

`!var`, `-var`, `2+var` also prints E030.

It seems that it can be fixed by adding at least `!` (and others?) to `checkPunctuators()` call in https://github.com/jshint/jshint/blob/master/src/jshint.js#L2246. Is it a good idea to send pull request - or maybe you plan to get rid of such workaround soon? Thanks.

**Comment by jugglinmike** [comment]
Thanks for the report! We've fixed this in [the feature branch for the next minor version of JSHint, v2.10](https://github.com/jshint/jshint/tree/v2.10.0) (commit 8b1bbf38cfe4690e810db6c7bbc591575abffbfe at the time of this writing, though that SHA may change by the time we release). I'll mark this report with the "Has PR" label to help communicate this (in truth, we're not quite ready to release version 2.10.0, but I'm hoping this will help contributors avoid duplicating work that's already been done).

I know that doesn't help you today, though. In the short term, the best I can offer is an unsightly workaround: since JSHint 2.9.5 expects an identifier, create a temporary variable to store the value in the conditional expression. For your specific example, that could look something like this:

```diff
-console.log([...+process.env.TEST ? [1] : [], 2]);
+const value=+process.env.TEST;
+console.log([...value ? [1] : [], 2]);
```

If you'd rather limit the definition of such temporary bindings, you could also wrap that whole thing in a block:

```diff
-console.log([...+process.env.TEST ? [1] : [], 2]);
+{
+  const value=+process.env.TEST;
+  console.log([...value ? [1] : [], 2]);
+}
```

Hope this helps!

**Comment by jugglinmike** [comment]
Good news! This has been resolved in [JSHint version 2.10.0](https://jshint.com/blog/2019-02-05/release-2-10-0/) released earlier this week.

    $ ./bin/jshint --version
    jshint v2.10.1
    $ echo '/* jshint esversion: 6 */ console.log([...+process.env.TEST ? [1] : [], 2]);' | ./bin/jshint -
    $ echo $?
    0
